### PR TITLE
issues-174 Add support for WINDOWS statement

### PR DESCRIPTION
**Pull request by ikrivosheev**
## PR Info

- Closes https://github.com/SeaQL/sea-query/issues/174

## Adds

Add support for WINDOW statement

**TODO:**

- [x] support window statement
- [x] add more helpers functions
- [x] add more tests
- [x] add more docs

### CODE REVIEW

File: src/backend/query_builder.rs [hunk 1]
```diff
@@ -1177,6 +1198,70 @@ pub trait QueryBuilder: QuotedBuilder {
...
+            });
+        }
+
+        if let Some(frame) = &window.frame {
+            match frame.r#type {
+                FrameType::Range => write!(sql, " RANGE ").unwrap(),
+                FrameType::Rows => write!(sql, " ROWS ").unwrap(),
+            };
+            if let Some(end) = &frame.end {
+                write!(sql, " BETWEEN ").unwrap();
```
Comment (billy1624) [inline src/backend/query_builder.rs:1255]:
```suggestion
                write!(sql, "BETWEEN ").unwrap();
```

I guess we have an excessive space before `BETWEEN`?

File: src/backend/query_builder.rs [hunk 2]
```diff
@@ -1177,6 +1198,70 @@ pub trait QueryBuilder: QuotedBuilder {
...
+    /// Translate [`WindowStatement`] into SQL statement.
+    fn prepare_window_statement(
+        &self,
+        window: &WindowStatement,
+        sql: &mut SqlWriter,
+        collector: &mut dyn FnMut(Value),
+    ) {
+        let (first, exprs) = &window.partition_by;
+        write!(sql, " PARTITION BY ").unwrap();
+        self.prepare_simple_expr(first, sql, collector);
```
Comment (billy1624) [inline src/backend/query_builder.rs:1229]: From what I understand the `PARTITION BY` is not mandatory?
For instance, we can write
```sql
SELECT salary, sum(salary) OVER (ORDER BY salary) FROM empsalary;
```
Please correct me if I'm wrong

**Comment by billy1624** [comment]
Hey @ikrivosheev, thank you so much for the contributions!

REF:
- https://dev.mysql.com/doc/refman/8.0/en/window-function-descriptions.html
- https://www.sqlite.org/windowfunctions.html
- https://www.postgresql.org/docs/current/tutorial-window.html

File: src/backend/query_builder.rs (hunk 2 above)
Comment (ikrivosheev) [reply src/backend/query_builder.rs]: Yes, you are right. Done

**Comment by ikrivosheev** [comment]
@billy1624 I have done work with WINDOW functions)

File: Cargo.toml [hunk 3]
```diff
@@ -36,6 +36,7 @@ name = "sea_query"
 path = "src/lib.rs"
 
 [dependencies]
+either = "1.6.1"
```
[the remark this answers is not in the document]
Comment (ikrivosheev) [reply Cargo.toml]: Thank you for review. Done

File: src/query/select.rs [hunk 4]
```diff
@@ -64,12 +63,20 @@ pub enum SelectDistinct {
...
 }
 
+/// Window type in [`SelectExpr`]
+pub enum WindowSelectType {
+    /// Name in [`SelectStatement`]
+    Name(DynIden),
+    /// Inline query in [`SelectExpr`]
+    Query(WindowStatement),
+}
+
```
Comment (tyt2y3) [inline src/query/select.rs:74]: I like this. The type explains iteself

**Review by tyt2y3 — APPROVED** [review]

**Comment by tyt2y3** [comment]
I think at a minimum sea-orm only needs to update the sea-query dependency (after we released a new sea-query version) and these features will be exposed in sea-orm.

**Pull request merged**